File: sdks/python/src/agentue/runner/redis_bridge.py
"""Redis Streams event bridge and durable heartbeat index."""

from __future__ import annotations

import time
from collections.abc import Awaitable, Mapping
from dataclasses import dataclass
from typing import Any, Protocol

from agentue.runner.types import RunnerOptions


class AsyncRedisCommands(Protocol):
    """The Redis command subset used by Runner."""

    def delete(self, *names: str) -> Awaitable[Any]: ...

    def expire(self, name: str, seconds: int) -> Awaitable[Any]: ...

    def hget(self, name: str, key: str) -> Awaitable[Any]: ...

    def hset(self, name: str, *args: Any, **kwargs: Any) -> Awaitable[Any]: ...

    def set(self, name: str, value: str, **kwargs: Any) -> Awaitable[Any]: ...

    def xadd(self, name: str, fields: Mapping[str, str]) -> Awaitable[Any]: ...

    def xread(self, streams: Mapping[str, str], **kwargs: Any) -> Awaitable[Any]: ...

    def zadd(self, name: str, mapping: Mapping[str, float]) -> Awaitable[Any]: ...

    def zrangebyscore(self, name: str, minimum: float | str, maximum: float | str) -> Awaitable[Any]: ...

    def zrem(self, name: str, *values: str) -> Awaitable[Any]: ...


@dataclass(frozen=True, slots=True)
class StoredEvent:
    cursor: str
    data: str


class RedisEventBridge:
    """Persist Runner events and liveness without owning the Redis client."""

    def __init__(
        self,
        redis: AsyncRedisCommands,
        options: RunnerOptions,
        *,
        wall_clock: Any = time.time,
    ) -> None:
        self._redis = redis
        self._options = options
        self._wall_clock = wall_clock

    async def initialize(self, task_id: str, initial_model: str) -> None:
        await self._redis.hset(
            self._state_key(task_id),
            mapping={"status": "running", "last_seq": "0", "initial_model": initial_model},
        )
        await self._redis.expire(self._state_key(task_id), self._options.task_ttl)
        await self.touch(task_id)

    async def publish(self, task_id: str, event_json: str, seq: int) -> str:
        event_id = await self._redis.xadd(self._stream_key(task_id), {"event": event_json})
        await self._redis.expire(self._stream_key(task_id), self._options.task_ttl)
        await self._redis.hset(self._state_key(task_id), mapping={"last_seq": str(seq)})
        return _text(event_id)

    async def read(self, task_id: str, cursor: str) -> list[StoredEvent]:
        response = await self._redis.xread(
            {self._stream_key(task_id): cursor},
            count=self._options.stream_read_count,
            block=self._options.stream_block_ms,
        )
        events: list[StoredEvent] = []
        for _, messages in response or []:
            for event_id, fields in messages:
                payload = fields.get("event") if isinstance(fields, dict) else None
                if payload is None and isinstance(fields, dict):
                    payload = fields.get(b"event")
                if payload is None:
                    continue
                events.append(StoredEvent(cursor=_text(event_id), data=_text(payload)))
        return events

    async def touch(self, task_id: str) -> None:
        await self._redis.zadd(self._heartbeat_key, {task_id: self._wall_clock()})

    async def expired_task_ids(self) -> list[str]:
        cutoff = self._wall_clock() - self._options.heartbeat_timeout
        values = await self._redis.zrangebyscore(self._heartbeat_key, "-inf", cutoff)
        return [_text(value) for value in values]

    async def claim_recovery(self, task_id: str) -> bool:
        claimed = await self._redis.set(
            self._claim_key(task_id),
            "1",
            nx=True,
            ex=self._options.recovery_claim_ttl,
        )
        return bool(claimed)

    async def release_recovery(self, task_id: str) -> None:
        await self._redis.delete(self._claim_key(task_id))

    async def status(self, task_id: str) -> str | None:
        value = await self._redis.hget(self._state_key(task_id), "status")
        return _text(value) if value is not None else None

    async def last_seq(self, task_id: str) -> int:
        value = await self._redis.hget(self._state_key(task_id), "last_seq")
        return int(_text(value)) if value is not None else 0

    async def initial_model(self, task_id: str) -> str | None:
        value = await self._redis.hget(self._state_key(task_id), "initial_model")
        return _text(value) if value is not None else None

    async def mark_terminal(self, task_id: str, status: str) -> None:
        await self._redis.hset(self._state_key(task_id), mapping={"status": status})
        await self._redis.expire(self._state_key(task_id), self._options.task_ttl)
        await self._redis.zrem(self._heartbeat_key, task_id)

    async def forget_heartbeat(self, task_id: str) -> None:
        await self._redis.zrem(self._heartbeat_key, task_id)

    @property
    def _heartbeat_key(self) -> str:
        return f"{self._options.key_prefix}:heartbeats"

    def _stream_key(self, task_id: str) -> str:
        return f"{self._options.key_prefix}:{task_id}:events"

    def _state_key(self, task_id: str) -> str:
        return f"{self._options.key_prefix}:{task_id}:state"

    def _claim_key(self, task_id: str) -> str:
        return f"{self._options.key_prefix}:{task_id}:recovery-claim"


def _text(value: Any) -> str:
    if isinstance(value, bytes):
        return value.decode("utf-8")
    return str(value)

File: sdks/go/runner/redis_bridge.go
package runner

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"strconv"
	"strings"
	"time"

	"github.com/redis/go-redis/v9"
)

const (
	defaultKeyPrefix = "agentue:runner"
	defaultTaskTTL   = 24 * time.Hour
	defaultReadBlock = time.Second
	defaultReadCount = 100
)

var initializeScript = redis.NewScript(`
if redis.call('EXISTS', KEYS[1]) == 1 then
  return redis.error_reply('AGENTUE_CONFLICT')
end
local cursor = redis.call('XADD', KEYS[2], '*', 'event', ARGV[4])
redis.call('HSET', KEYS[1],
  'status', ARGV[1],
  'last_seq', ARGV[2],
  'initial_model', ARGV[3],
  'last_event', ARGV[4],
  'last_cursor', cursor)
redis.call('EXPIRE', KEYS[1], ARGV[5])
redis.call('EXPIRE', KEYS[2], ARGV[5])
return cursor
`)

var publishScript = redis.NewScript(`
local status = redis.call('HGET', KEYS[1], 'status')
if not status then
  return redis.error_reply('AGENTUE_NOT_FOUND')
end
local last_seq = tonumber(redis.call('HGET', KEYS[1], 'last_seq'))
local seq = tonumber(ARGV[2])
if seq == last_seq and redis.call('HGET', KEYS[1], 'last_event') == ARGV[3] then
  return redis.call('HGET', KEYS[1], 'last_cursor')
end
if status ~= ARGV[1] then
  return redis.error_reply('AGENTUE_CONFLICT')
end
if seq <= last_seq then
  return redis.error_reply('AGENTUE_CONFLICT')
end
local cursor = redis.call('XADD', KEYS[2], '*', 'event', ARGV[3])
redis.call('HSET', KEYS[1], 'last_seq', ARGV[2], 'last_event', ARGV[3], 'last_cursor', cursor)
redis.call('EXPIRE', KEYS[1], ARGV[4])
redis.call('EXPIRE', KEYS[2], ARGV[4])
return cursor
`)

var markTerminalScript = redis.NewScript(`
local current = redis.call('HGET', KEYS[1], 'status')
if not current then
  return redis.error_reply('AGENTUE_NOT_FOUND')
end
if current ~= ARGV[1] and current ~= ARGV[2] then
  return redis.error_reply('AGENTUE_CONFLICT')
end
redis.call('HSET', KEYS[1], 'status', ARGV[2])
redis.call('EXPIRE', KEYS[1], ARGV[3])
redis.call('EXPIRE', KEYS[2], ARGV[3])
return 'OK'
`)

// RedisEventBridge persists Runner events and liveness without owning the Redis client.
type RedisEventBridge struct {
	client  redis.UniversalClient
	options BridgeOptions
}

func NewRedisEventBridge(client redis.UniversalClient, options BridgeOptions) *RedisEventBridge {
	if options.KeyPrefix == "" {
		options.KeyPrefix = defaultKeyPrefix
	}
	if options.TaskTTL <= 0 {
		options.TaskTTL = defaultTaskTTL
	}
	if options.ReadBlock <= 0 {
		options.ReadBlock = defaultReadBlock
	}
	if options.ReadCount <= 0 {
		options.ReadCount = defaultReadCount
	}
	return &RedisEventBridge{client: client, options: options}
}

func (bridge *RedisEventBridge) Initialize(
	ctx context.Context,
	taskID string,
	initialModel json.RawMessage,
	startEvent json.RawMessage,
	seq uint64,
) error {
	_, err := initializeScript.Run(ctx, bridge.client,
		[]string{bridge.stateKey(taskID), bridge.streamKey(taskID)},
		string(StatusRunning),
		seq,
		string(initialModel),
		string(startEvent),
		bridge.ttlSeconds(),
	).Result()
	if err != nil {
		if containsRedisError(err, "AGENTUE_CONFLICT") {
			return fmt.Errorf("%w: task %q already exists", ErrConflict, taskID)
		}
		return fmt.Errorf("initialize event stream %q: %w", taskID, err)
	}
	return nil
}

func (bridge *RedisEventBridge) Delete(ctx context.Context, taskID string) error {
	if err := bridge.client.Del(ctx, bridge.stateKey(taskID), bridge.streamKey(taskID)).Err(); err != nil {
		return fmt.Errorf("delete event stream %q: %w", taskID, err)
	}
	return nil
}

func (bridge *RedisEventBridge) State(ctx context.Context, taskID string) (State, error) {
	values, err := bridge.client.HGetAll(ctx, bridge.stateKey(taskID)).Result()
	if err != nil {
		return State{}, fmt.Errorf("read event stream %q: %w", taskID, err)
	}
	if len(values) == 0 {
		return State{}, fmt.Errorf("%w: task %q", ErrNotFound, taskID)
	}
	lastSeq, err := strconv.ParseUint(values["last_seq"], 10, 64)
	if err != nil {
		return State{}, fmt.Errorf("decode event stream %q sequence: %w", taskID, err)
	}
	return State{
		TaskID: taskID, Status: Status(values["status"]), LastSeq: lastSeq,
		LastCursor: values["last_cursor"], InitialModel: json.RawMessage(values["initial_model"]),
	}, nil
}

func (bridge *RedisEventBridge) Publish(
	ctx context.Context,
	taskID string,
	event json.RawMessage,
	seq uint64,
) (string, error) {
	value, err := publishScript.Run(ctx, bridge.client,
		[]string{bridge.stateKey(taskID), bridge.streamKey(taskID)},
		string(StatusRunning),
		seq,
		string(event),
		bridge.ttlSeconds(),
	).Result()
	if err != nil {
		switch {
		case containsRedisError(err, "AGENTUE_NOT_FOUND"):
			return "", fmt.Errorf("%w: task %q", ErrNotFound, taskID)
		case containsRedisError(err, "AGENTUE_CONFLICT"):
			return "", fmt.Errorf("%w: task %q sequence %d", ErrConflict, taskID, seq)
		default:
			return "", fmt.Errorf("publish event for task %q: %w", taskID, err)
		}
	}
	return text(value), nil
}

func (bridge *RedisEventBridge) EventsThrough(ctx context.Context, taskID, end string) ([]StoredEvent, error) {
	if end == "" {
		end = "+"
	}
	values, err := bridge.client.XRange(ctx, bridge.streamKey(taskID), "-", end).Result()
	if err != nil {
		return nil, fmt.Errorf("range events for task %q: %w", taskID, err)
	}
	return storedEvents(values), nil
}

func (bridge *RedisEventBridge) Read(ctx context.Context, taskID, after string) ([]StoredEvent, error) {
	if after == "" {
		after = "0-0"
	}
	values, err := bridge.client.XRead(ctx, &redis.XReadArgs{
		Streams: []string{bridge.streamKey(taskID), after},
		Count:   bridge.options.ReadCount, Block: bridge.options.ReadBlock,
	}).Result()
	if errors.Is(err, redis.Nil) {
		return nil, nil
	}
	if err != nil {
		return nil, fmt.Errorf("read events for task %q: %w", taskID, err)
	}
	var result []StoredEvent
	for _, stream := range values {
		result = append(result, storedEvents(stream.Messages)...)
	}
	return result, nil
}

func (bridge *RedisEventBridge) MarkTerminal(ctx context.Context, taskID string, status Status) error {
	if !status.Terminal() {
		return fmt.Errorf("invalid terminal status %q", status)
	}
	_, err := markTerminalScript.Run(ctx, bridge.client,
		[]string{bridge.stateKey(taskID), bridge.streamKey(taskID)},
		string(StatusRunning),
		string(status),
		bridge.ttlSeconds(),
	).Result()
	if err != nil {
		switch {
		case containsRedisError(err, "AGENTUE_NOT_FOUND"):
			return fmt.Errorf("%w: task %q", ErrNotFound, taskID)
		case containsRedisError(err, "AGENTUE_CONFLICT"):
			return fmt.Errorf("%w: task %q is already terminal", ErrConflict, taskID)
		default:
			return fmt.Errorf("mark task %q terminal: %w", taskID, err)
		}
	}
	return nil
}

func (bridge *RedisEventBridge) stateKey(taskID string) string {
	return bridge.options.KeyPrefix + ":" + taskID + ":state"
}

func (bridge *RedisEventBridge) streamKey(taskID string) string {
	return bridge.options.KeyPrefix + ":" + taskID + ":events"
}

func (bridge *RedisEventBridge) ttlSeconds() int64 {
	seconds := int64(bridge.options.TaskTTL / time.Second)
	if seconds < 1 {
		return 1
	}
	return seconds
}

func storedEvents(values []redis.XMessage) []StoredEvent {
	result := make([]StoredEvent, 0, len(values))
	for _, value := range values {
		data, exists := value.Values["event"]
		if !exists {
			continue
		}
		result = append(result, StoredEvent{Cursor: value.ID, Data: json.RawMessage(text(data))})
	}
	return result
}

func text(value any) string {
	switch value := value.(type) {
	case string:
		return value
	case []byte:
		return string(value)
	default:
		return fmt.Sprint(value)
	}
}

func containsRedisError(err error, value string) bool {
	return err != nil && strings.Contains(err.Error(), value)
}
